### PR TITLE
Vectorized Convert Rotations Node

### PR DESCRIPTION
**Pull request by kdsin23**
finished my code about this request

### CODE REVIEW

**Comment by OmarEmaraDev** [comment]
Can you confirm that the list computations are identical to the single computations we had before?

**Comment by kdsin23** [comment]
> Haven't looked at the code yet, but can you confirm that the list computations are identical to the single computations we had before?

yes, sure. i modified the code accroding to you and Kuldeep Singh suggestions last time.

File: animation_nodes/math/conversion.pyx [hunk 1]
```diff
@@ -120,7 +122,20 @@ cdef toPyEuler3(Euler3* e):
...
+    cdef float sx = sin(e.x * 0.5)
+    cdef float cy = cos(e.y * 0.5)
+    cdef float sy = sin(e.y * 0.5)
+    cdef float cz = cos(e.z * 0.5)
+    cdef float sz = sin(e.z * 0.5)
+    
+    q.w = cx * cy * cz + sx * sy * sz
+    q.x = sx * cy * cz - cx * sy * sz
+    q.y = cx * sy * cz + sx * cy * sz
+    q.z = cx * cy * sz - sx * sy * cz
```
Comment (OmarEmaraDev) [inline animation_nodes/math/conversion.pyx:138]: Looks like you have some common expressions here, try to group them. This applies to the rest of the functions as well.

File: animation_nodes/math/conversion.pyx [hunk 2]
```diff
@@ -139,6 +154,48 @@ cdef setQuaternion(Quaternion* q, value):
 cdef toPyQuaternion(Quaternion* q):
     return PyQuaternion((q.w, q.x, q.y, q.z))
 
+#https://www.euclideanspace.com/maths/geometry/rotations/conversions/quaternionToMatrix/index.htm
```
Comment (OmarEmaraDev) [inline animation_nodes/math/conversion.pyx:157]: Put a space after the `#`, applies to rest of comments.

File: animation_nodes/math/conversion.pyx [hunk 3]
```diff
@@ -139,6 +154,48 @@ cdef setQuaternion(Quaternion* q, value):
...
+cdef quaternionToMatrix4(Matrix4 *m, Quaternion *q):
+    quaternionNormalize_InPlace(q)
+    cdef float w, x, y, z
+    w, x, y, z = q.w, q.x, q.y, q.z
+    cdef float ww = w * w
+    cdef float xx = x * x
+    cdef float yy = y * y
+    cdef float zz = z * z
+
+    cdef invs = 1 / (xx + yy + zz +ww)
```
Comment (OmarEmaraDev) [inline animation_nodes/math/conversion.pyx:167]: Space between `+ww`.

File: animation_nodes/math/conversion.pyx [hunk 4]
```diff
@@ -139,6 +154,48 @@ cdef setQuaternion(Quaternion* q, value):
...
+#https://en.wikipedia.org/wiki/Conversion_between_quaternions_and_Euler_angles
+cdef quaternionToEuler3(Euler3 *e, Quaternion *q):
+    quaternionNormalize_InPlace(q)
+    cdef float sinrCosp = 2 * (q.w * q.x + q.y * q.z)
+    cdef float cosrCosp = 1 - 2 * (q.x * q.x + q.y * q.y)
+    e.x = atan2(sinrCosp, cosrCosp)
+
+    cdef float sinp = 2 * (q.w * q.y - q.z * q.x)
+    if abs(sinp) >= 1.0:
+        e.y = copysign(PI/2, sinp)
```
Comment (OmarEmaraDev) [inline animation_nodes/math/conversion.pyx:191]: Spaces.

File: animation_nodes/math/quaternion.pyx [hunk 5]
```diff
@@ -49,6 +49,17 @@ cdef void quatFromAxisAngle(Quaternion *q, Vector3 *axis, float angle):
     q.z = axis.z * factor
     q.w = cq
 
+#https://www.euclideanspace.com/maths/geometry/rotations/conversions/quaternionToAngle/index.htm
+cdef void quatToAxisAngle(Vector3 *v, float *a, Quaternion *q):
+    quaternionNormalize_InPlace(q)
+    cdef float k = sqrt(1 - q.w * q.w)
+    
+    v.x = q.x / k
```
Comment (OmarEmaraDev) [inline animation_nodes/math/quaternion.pyx:57]: When you have division like this, insert `@cython.cdivision(True)` and make sure you don't have zero division.

File: animation_nodes/nodes/rotation/convert_rotations.py [hunk 6]
```diff
@@ -70,26 +91,66 @@ def drawLabel(self):
...
+                    return "quaternion = Quaternion(axis, angle)"
 
     def getUsedModules(self):
         return ["math"]
+
+    def matricesToQuaternions(self, matrices):
+        return Matrix4x4List.toQuaternions(matrices)
+
+    def matricesToEulers(self, matrices):
+        return Matrix4x4List.toEulers(matrices)
```
Comment (OmarEmaraDev) [inline animation_nodes/nodes/rotation/convert_rotations.py:156]: No need to define those. Embed the calls directly to the execution code.

**Review by OmarEmaraDev — CHANGES_REQUESTED** [review]
Move all of those functions to `animation_nodes/math/rotation_conversion`.

File: animation_nodes/math/conversion.pyx (hunk 1 above)
Comment (kdsin23) [reply animation_nodes/math/conversion.pyx:138]: sorry, i don't fully understand how to group them. i guess like this:
`cdef float cxcy = cx * cy`
 `cdef float sxcy = sx * cy`
 `cdef float cxsy = cx * sy`
` cdef float sxsy = sx * sy`

`q.w = cxcy * cz + sxsy *sz`
`q.x =  sxcy * cz - cxsy * sz`
`q.y = cxsy * cz + sxsy * sz`
`q.z = cxcy * sz -  sxsy *cz`

Comment (OmarEmaraDev) [reply animation_nodes/math/conversion.pyx:138]: @kdsin23 Yes, exactly.

Comment (kdsin23) [reply animation_nodes/math/conversion.pyx:138]: Thank you!

Comment (kdsin23) [reply animation_nodes/math/conversion.pyx:138]: BTW, is there a handy way to move all of those functions to `animation_nodes/math/rotation_conversion`?

Comment (OmarEmaraDev) [reply animation_nodes/math/conversion.pyx:138]: Handy as in automatic? Not that I know of.

File: animation_nodes/nodes/rotation/convert_rotations.py (hunk 6 above)
Comment (kdsin23) [reply animation_nodes/nodes/rotation/convert_rotations.py]: i don't know why my CHANG.MD file alway has a conflicts?

**Comment by kdsin23** [comment]
Error compiling Cython file:
------------------------------------------------------------
...
    c.b = value[2]
    c.a = value[3]

cdef toPyColor(Color* c):
    return PyColor((c.r, c.g, c.b, c.a)
^
------------------------------------------------------------

animation_nodes/math/conversion.pyx:161:0: Expected ')', found 'DEDENT'
Traceback (most recent call last):
  File "setup.py", line 236, in <module>
    main()
  File "setup.py", line 70, in main
    main_Build(args[1:], configs)
  File "setup.py", line 153, in main_Build
    changedFileStates = build(configs, skipCompilation = "--nocompile" in options)
  File "/home/runner/work/animation_nodes/animation_nodes/_setuputils/generic.py", line 241, in wrapper
    function(*args, **kwargs)
  File "setup.py", line 192, in build
    execute_Cythonize(setupInfoList, addonDirectory, configs)
  File "/home/runner/work/animation_nodes/animation_nodes/_setuputils/cythonize.py", line 8, in execute_Cythonize
    task.execute()
  File "/home/runner/work/animation_nodes/animation_nodes/_setuputils/cythonize.py", line 29, in execute
    cythonize(self.path, include_path = self.includePaths, compiler_directives = {"language_level" : "3"})
  File "/opt/hostedtoolcache/Python/3.8.6/x64/lib/python3.8/site-packages/Cython/Build/Dependencies.py", line 1102, in cythonize
    cythonize_one(*args)
  File "/opt/hostedtoolcache/Python/3.8.6/x64/lib/python3.8/site-packages/Cython/Build/Dependencies.py", line 1225, in cythonize_one
    raise CompileError(None, pyx_file)
Cython.Compiler.Errors.CompileError: /home/runner/work/animation_nodes/animation_nodes/animation_nodes/math/conversion.pyx
Error: Process completed with exit code 1.

how to fix the issue?

**Comment by kdsin23** [comment]
long number

why has the long long number in first matrix?

File: animation_nodes/nodes/rotation/convert_rotations.py (hunk 6 above)
Comment (kdsin23) [reply animation_nodes/nodes/rotation/convert_rotations.py]: so i keep the original file, it is no conflicts now

**Comment by OmarEmaraDev** [comment]
@kdsin23 Looks like you don't initialize the last row and column of the matrix. `[n][3]` and `[3][n]` for all `n` should be initialized to zero, except `[3][3]` which should be initialized to 1.

**Comment by OmarEmaraDev** [comment]
@kdsin23 I am guessing you still don't have a local development setup. You should really do that, things will break and you will not be able to fix them on Github. I can see you are already having difficulties.

**Comment by kdsin23** [comment]
> @kdsin23 Looks like you don't initialize the last row and column of the matrix. `[n][3]` and `[3][n]` for all `n` should be initialized to zero, except `[3][3]` which should be initialized to 1.

how to initialize? i afraid i can not

**Comment by kdsin23** [comment]
> @kdsin23 I am guessing you still don't have a local development setup. You should really do that, things will break and you will not be able to fix them on Github. I can see you are already having difficulties.

sorry, i was always failing to use git in my computer, i don't know why.

**Comment by OmarEmaraDev** [comment]
@kdsin23  Initialize the matrix like this:

```
m.a44 = 1.0
m.a14 = m.a24 = m.a34 = 0.0
m.a41 = m.a42 = m.a43 = 0.0
```

Are you on Window or on Linux? What are the steps you took to work with the repository?

**Comment by kdsin23** [comment]
> @kdsin23 Initialize the matrix like this:
> 
> ```
> m.a44 = 1.0
> m.a14 = m.a24 = m.a34 = 0.0
> m.a41 = m.a42 = m.a43 = 0.0
> ```
> 
> Are you on Window or on Linux? What are the steps you took to work with the repository?

Thank you!  i am on Linux. why can not create a new branch for me in git remote.

**Comment by OmarEmaraDev** [comment]
@kdsin23 

# Setup Repository

To setup the repository for the very first time time.

```
# Clone your fork.
git clone https://github.com/kdsin23/animation_nodes.git
# Add Jacques' repository as a remote.
git remote add upstream https://github.com/JacquesLucke/animation_nodes.git
```

# Update Repository

To update the master branch before you start working on anything.

```
# Change active branch to master.
git checkout master
# Fetch changes from upstream.
git fetch upstream
# Merge upstream master to local master.
git merge upstream/master
```

# Create A New Patch

To create a new patch for Animation Nodes.
Update the repository first as described above.

```
# Change active branch to master.
git checkout master
# Create a new branch.
git checkout -b your-branch-name

# Make your changes ...

# Track the changes.
git add .
# Commit changes.
git commit -m "Your commit message here."
# Push the branch.
git push --set-upstream origin your-branch-name
```

The create the pull request on Github.

# Fix Conflicts

When there are conflicts, this is what you do.
Update the repository first as described above.

```
# Change active branch to your branch
git checkout your-branch-name
# Merge master
git merge master

# Git will tell you there are conflicts and the merge
# couldn't be completed. Fix those conflicts with
# whatever method you prefer.

# Complete Merge
git commit
# Push Changes
git push
```

**Comment by kdsin23** [comment]
Thank you for so detail steps. 

BTW, this request still needs to be modified or not? i feel it's almost has been finished.

**Comment by OmarEmaraDev** [comment]
@kdsin23 There are still conflicts that you need to solve. Moreover, I can see some changes in `math/quaternion` those need to move to rotation conversion as well.

This pull request is coming from a branch called `patch-1`, so use the methods I described using ` patch-1` as "your branch name".

**Comment by kdsin23** [comment]
> @kdsin23 There are still conflicts that you need to solve. Moreover, I can see some changes in `math/quaternion` those need to move to rotation conversion as well.
> 
> This pull request is coming from a branch called `patch-1`, so use the methods I described using ` patch-1` as "your branch name".

Thank you!i found in the `math/quaternion` there is a original function `quatFromAxisAngle`, so is it necessary to move the `quatToAxisAngle`, i feel this two functions are pairs. what do you think?

**Comment by OmarEmaraDev** [comment]
@kdsin23 Yes, move that function as well if necessary.

**Comment by kdsin23** [comment]
> @kdsin23 Yes, move that function as well if necessary.

OK,i will do that

**Comment by kdsin23** [comment]
sorry , i don't know why appears this!

```
To prevent you from losing history, non-fast-forward updates were rejected
Merge the remote changes before pushing again.  See the 'Note about
fast-forwards' section of 'git push --help' for details.
```

**Comment by OmarEmaraDev** [comment]
@kdsin23 Did you merge extra nodes into your master at some point? Because it looks like you did.
Generally, you should never commit something to your master, your master should be identical to ours.

We need to reset your master. Update upstream as explained above, then:

```
git checkout master
git reset --hard upstream/master
git push origin master --force
```

Then you should also reset `patch-1` and commit again. I don't want to complicate things for you, so if you want to create a new pull request with a new branch. You should follow my instructions above to create the patch after you reset master.

**Comment by kdsin23** [comment]
> @kdsin23 Did you merge extra nodes into your master at some point? Because it looks like you did.
> Generally, you should never commit something to your master, your master should be identical to ours.
> 
> We need to reset your master. Update upstream as explained above, then:
> 
> ```
> git checkout master
> git reset --hard upstream/master
> git push origin master --force

sorry, i feel the patch-1 is kinda  messy, so i pull request new branch.

**Pull request closed without merging**